.github/workflows/ci-workflow.yml

```diff
@@ -67,4 +68,28 @@ jobs:
           SLACK_MESSAGE: ':man-shrugging:'
           SLACK_TITLE: Build Cancelled
           SLACK_USERNAME: rtCamp
-          SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
\ No newline at end of file
+          SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
+
+      - name: Login to Docker
+        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      
+      - name: Login to Docker for Heroku
+        run: docker login --username=_ --password=${{ secrets.HEROKU_API_KEY }} registry.heroku.com
+
+      - name: Build image
+        run: docker build --target production -t ${{ secrets.DOCKER_USERNAME }}/todo-list_production:latest -t ${{ secrets.DOCKER_USERNAME }}/todo-list_production:${GITHUB_REF#refs/heads/}.${{ github.sha }} .
+      
+      - name: Push to Docker
+        run: docker push ${{ secrets.DOCKER_USERNAME }}/todo-list_production --all-tags
+
+      - name: Tag for Heroku
+        run: docker tag ${{ secrets.DOCKER_USERNAME }}/todo-list_production:latest registry.heroku.com/rich-todo-list/web
+      
+      - name: Push to heroku
+        run: docker push registry.heroku.com/rich-todo-list/web
+
+      - name: Deploy heroku
+        run: heroku container:release web --app rich-todo-list
```

Comment:
You have added these steps onto the existing job. This means it will run whenever the tests pass for any branch.

We don't want that. The site hosted on Heroku would become too unstable when people are working on multiple branches. 

Move these into a second job that only runs for pushes to the master branch.

```
jobs:
  build:
    name: Build and test
    runs-on: ubuntu-latest
    env:
      ...
    steps:
      ...
  deploy:
    if: <fill me in>
    needs: build
    name: ...
    runs-on: ..
    env:
      ...
    steps:
       ...
```